### PR TITLE
client: report root's quota in statfs

### PR DESCRIPTION
**Pull request by jcsp**
When user is mounted a quota-restricted inode
as the root, report that inode's quota status
as the filesystem statistics in statfs.

This allows us to have a fairly convincing illusion
that someone has a filesystem to themselves, when
they're really mounting a restricted part of
the larger global filesystem.

Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -8779,15 +8763,51 @@ int Client::statfs(const char *path, struct statvfs *stbuf)
...
+
+  C_SaferCond cond;
+  objecter->get_fs_stats(stats, &cond);
+
+  client_lock.Unlock();
+  int rval = cond.wait();
+  client_lock.Lock();
+
+  if (cct->_conf->client_quota && root_ancestor
+      && root_ancestor->quota.max_bytes) {
```
Comment (ukernel) [inline src/client/Client.cc]: maybe we should use get_quota_root(root) instead of root_ancestor

**Comment by liewegas** [comment]
Should this be a config option maybe to control whether it is the quota vs full fs limit?  We can default to the quota one...

**Comment by gregsfortytwo** [comment]
Yeah, what Sage said.

Maybe even set it up as a server-side config that gets passed to the client on connect? Although it's not secure since the client still sees the whole cluster, it's more of a cluster admin decision than a user one.

File: src/client/Client.cc (hunk 1 above)
Comment (jcsp) [inline src/client/Client.cc]: What is the advantage to doing that?  It's not obvious to me

Comment (ukernel) [inline src/client/Client.cc]: I think root_ancestor->quota is filesystem root's quota. It's possible that the mounted sub-directory or its ancestor directories have quota

Comment (liewegas) [inline src/client/Client.cc]: I think that's what root_ancestor is?  Either the mounted inode, or the 
first ancestor that has a quota set.

**Comment by jcsp** [comment]
Superceded (by accident!) by https://github.com/ceph/ceph/pull/8747

**Pull request closed without merging**